lib/common/balancer/roundrobin.c

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2017 Justin Zhu
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#include "h2o/balancer.h"
+
+struct round_robin_t {
+    h2o_balancer_t super;
+    size_t next_pos;           /* counting next logic index */
+    size_t next_actual_target; /* indicate next actual target index */
+    size_t *floor_next_target; /* caching logic indices indicating next target should be used */
+    size_t pos_less_than;      /* return point for logic count */
+    pthread_mutex_t mutex;
+};
+
+static size_t selector(h2o_balancer_t *balancer, h2o_socketpool_target_vector_t *targets, int *tried)
+{
+    size_t i;
+    size_t result = 0;
+    struct round_robin_t *self = (void *)balancer;
+
+    pthread_mutex_lock(&self->mutex);
+
+    assert(targets->size != 0);
+    for (i = 0; i < targets->size; i++) {
```

Comment:
Do we need to be strict in retrying to connect to a backend that we failed to connect in the previous call to `h2o_socketpool_connect`?

If not, I think the algorithm can be made simpler to something like below for example (and as an added bonus we might be able to delete the `set_targets` callback).
```
struct round_robin_t {
    h2o_balancer_t super;
    size_t pos;
    uint16_t remaining_weight;
};

static void select_next(struct round_robin_t *self, h2o_socketpool_target_vector_t *targets)
{
    self->pos = (self->pos + 1) % targets->size;
    self->remaining_weight = targets[self->pos]->conf->weight;
}

static size_t selector(...)
{
    ...
    for (i = 0; i < targets->size; ++i) {
        if (!tried[self->pos]) {
            result = pos;
            if (--self->remaining_weight == 0)
                select_next(self, targets);
                self->pos = (self->pos + 1) % targets->size;
            return result;
        } else {
            select_next(self, targets);
        }
    }
    assert(!"unreachable");
}
```

Reply:
Thanks for the suggestion. Will make it simpler.